### ISSUE TITLE
How difficult would it be to have PSyclone read OpenACC directives?

### ISSUE THREAD

**Issue by joewallwork** [stfc/PSyclone#2275]
We have some code from a manual GPU port using OpenACC and are trying to replicate its performance using PSyclone.

If the code is formatted using PSyclone before the port then it is simple to take a diff between the manual port and PSyclone-generated port. However, our manual port was done using the original code, so its not so straightforward.

One option is just to re-apply the manual port to the reformatted code, although it is tedious and error-prone. I wonder if it might be possible to have PSyclone read the manual port code and reformat that, keeping the OpenACC directives in place?

Currently, running PSyclone without an optimisation script will remove all comments and reformat the code. I notice there is a [branch which keeps all comments](https://github.com/stfc/PSyclone/compare/1247_allow_comments). Does reading in the OpenACC sound like something achievable?

**Comment by arporter** [comment]
Hi @jwallwork23, I think it's doable but probably not easy. The keep-all-comments branch you've found was something I did for Victoria for precisely the same reason (with WavewatchIII). Having a quick look at it, it seems like the changes are all healthy ones apart from having to comment-out some sanity checks on the structure of the fparser2 parse tree.

It *should* be relatively straightforward to extend the frontend to handle OpenACC directives.

There is a hacky alternative which I think Victoria and Martin sometimes use - sed the ACC directives so that they are e.g. WRITE statements. These then get put into CodeBlocks by PSyclone and are preserved. Obviously, the results are not guaranteed to be syntatically correct (because of PSyclone's canonicalisation of loop/control structures) but it's a start.

**Comment by arporter** [comment]
> It _should_ be relatively straightforward to extend the frontend to handle OpenACC directives.

In fact, a first step would be simply to put such directives into CodeBlocks.

**Comment by jwallwork23** [comment]
Thanks for the info @arporter. I'll look into it in more detail and maybe we can discuss it sometime :)

**Comment by arporter** [comment]
Yes, I'd be happy to chat it through :-) I'm slightly concerned that allowing for comments will break a lot of things - I can't remember whether the test suite worked with the branch you found for instance.

**Comment by LonelyCat124** [comment]
@arporter Can we close this as completed or is this issue asking for directives to be converted to their PSyIR representation?

**Comment by arporter** [comment]
I think that it is only asking that we keep OpenACC directives and therefore I think it's done.
